Mentions / Content parsers

rbngzlv opened the pull request:
#### :tophat: What? Why?

Add mentions through content parsers.

I coded this feature with this goals in mind:
- The parsers can run inside a runner or outside it.
- Is possible that some parser will need access to an object to check some things.
- Now we have parsers for content, but in the future we can have another type of parsers.
- We are registering the content parsers in a in initializer, but perhaps, some models doesn't require to run all registered parsers, so a model can skip a parser:

```
class Comment < ApplicationRecord
  include Decidim::Parseable::Content
  parses_content :body, skip: "Decidim::Parsers::Content::MentionParser"
end
```

#### :pushpin: Related Issues

- Fixes: https://github.com/decidim/decidim-module-mentions/issues/1

#### :clipboard: Subtasks
- [X] Modify core to create the array of parsers and add a class to wrap these parsers so the API is the same for everyone
- [X] Add the mentions parser to core
- [X] Modify comments to parse the comments body before saving
- [x] Allow HTML in comment (React)
- [x] Sanitize user input
- [x] Add specs
- [ ] Add changelog

rbngzlv commented:
@decidim/lot-core I would greatly appreciate your comments on the structure of the code before continue.

mrcasals commented:
I like where this is going, but the structure is more complicated than what I had thought, but that doesn't mean it's bad! 😄

rbngzlv commented:
Yes, seems a bit complicated. Any constructive criticism will be welcome. I updated the pull request description to explain my goals on this.

josepjaume commented:
Hi! Nice work!

I'm not entirely sold on using concerns & callbacks for something like this. It feels like content parsing should be done at comment creation time, not every time it gets saved. Callbacks usually backfire as they don't have enough context on when to be triggered.

Moreover, we probably need a way to trigger actions after a comment is created, like notifying the mentioned user and so on. This probably calls for a different design.

What about having something like this instead:

```ruby
parse_data = Decidim::Parsers::Content::MentionParser.new(comment.body).parse
mentioned_users = parse_data.mentioned_users
comment.body = parse_data.rewrite
```

We could even `rewrite` at runtime (when returning the comment via GraphQL), so that we can change the way we construct the links in the future. This could be potentially done with a different class, something like a `Renderer` maybe?

I would personally create having multiple (registered) parsers in the future when we have played a bit with this interface and settled with a design - maybe when hashtags are introduced?

So, what do you think?

rbngzlv commented:
Hi @josepjaume!

Thanks for taking a look at this.

> Moreover, we probably need a way to trigger actions after a comment is created, like notifying the mentioned user and so on. This probably calls for a different design.

Yes, I was thinking about it, and I can't see a "clean" path to do it with the actual proposed structure.

> 
> What about having something like this instead:
> ```
> parse_data = Decidim::Parsers::MentionParser.new(comment.body)
> mentioned_users = parse_data.mentioned_users
> comment.body = parse_data.render
> ```

You would do this inside the `Decidim:: Comments:: CreateComment `, right?

> We could even rewrite at runtime (when returning the comment via GraphQL), so that we can change the way we construct the links in the future. This could be potentially done with a different class, something like a Renderer maybe?

This was something that had to be discussed, because the product team doesn't want to have dead links to the user profile if a user changes their nickname and this is the only way that I can see at the moment. Will be ok to have the comment saved with a token like `user:1` and replace this when rendering the comment?

> I would personally create having multiple (registered) parsers in the future when we have played a bit with this interface and settled with a design - maybe when hashtags are introduced?

@isaacmg410 is starting the work on hashtags, so we need to take a decision on this. Also, now that we are introducing HTML in the rendered comments, we will need a parser to remove HTML submitted by the user?

Perhaps we can have a short meeting on this?

/cc @mrcasals

josepjaume commented:
Hi @rbngzlv - I'll try to reply to all that's being said :D

About the "taking action after parsing" - I think getting that information on the `command` itself and acting upon it is enough. No need to create hooking mechanisms, pub/sub or anything like that, I think. But that might be something desirable on the future.

About the "dead links" issue: Don't think having a dead link here and there would be such an issue but anyway: What you propose seems fine to me. Having some special syntax to refer to entities that then is then parseable by a renderer looks like the way to go. What do you think about `@Entity:id`? We should be careful sanitizing this just before though, as this poses a risk if a user maliciously enters that input.

Does this look reasonable?

josepjaume commented:
To improve on the syntax I proposed, what do you think about having something like:

```ruby
# This class gets called on parsing time just before entering the content in the database.
class Decidim::ContentParsers::UserParser < Decidim::ContentParser
  # Gets initialized with the `content` from the comment. It can either receive
  # already rewritten content or regular content.
  def initialize(content); end;

  # Rewrites the content by substituting user mentions by @user:id.
  # This will get saved to the database.
  def rewrite; end

  # Returns a hash with metadata (potentially the found entities). This metadata is
  # accessible at parsing time so it can be acted upon (sending emails to the users)
  # or maybe even stored at the DB for later consultation.
  def metadata; end;
end

# This class gets used at rendering time by the backend.
class Decidim::ContentParsers::UserRenderer < Decidim::ContentRenderer
  include Rails.application.routes #or whatever you need

  def initialize(content); end;
  
  # Replaces any @user:id with a link to the profile.
  def render(content); end;
end
```

oriolgual commented:
I also like the `@Entity:id` idea, we could even use Rails' GlobalId

rbngzlv commented:
@josepjaume just this morning I was thinking that the "parser runner" needs to return a hash with statistics so we can do actions based on this (like notifications), so I agree with your design.

I'll start working on this tomorrow

rbngzlv commented:
@decidim/lot-core I pushed the new content processor structure and now I need guidance for implementing the graphql/react part:

- Where I can process the comments body before they are returned to the browser?

From now we will be putting HTML in the rendered comments, so:

- We need to sanitize the user input before render it or before save it to the database? I think that we can do it just before apply the content renderers.
- In React we need to use `dangerouslySetInnerHTML` (https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml) ?

mrcasals commented:
@rbngzlv you could add a new field on the GraphQL type to call a method on the model, for example, which delegates the work to a renderer

rbngzlv commented:
Pull request ready for review.

/cc @decidim/lot-core

mrcasals commented on decidim-comments/config/locales/en.yml:
```diff
@@ -14,6 +14,11 @@ en:
             email_subject: There is a new reply of your comment from %{author_name} in <a href="%{resource_url}">%{resource_title}</a>
             notification_title: There is a new reply of your comment from <a href="%{author_path}">%{author_name} %{author_nickname}</a> in <a href="%{resource_path}">%{resource_title}</a>
+        user_mentioned:
+          email_intro: You has been mentioned
```
You

mrcasals replied:
You have

mrcasals replied:
[You have been](https://www.youtube.com/watch?v=W3q8Od5qJio)

mrcasals commented on decidim-comments/config/locales/en.yml:
```diff
@@ -14,6 +14,11 @@ en:
             notification_title: There is a new reply of your comment from <a href="%{author_path}">%{author_name} %{author_nickname}</a> in <a href="%{resource_path}">%{resource_title}</a>
+        user_mentioned:
+          email_intro: You has been mentioned
+          email_outro: You have received this notification because you has been mentioned in "%{resource_title}".
```
You have

mrcasals commented on decidim-comments/config/locales/en.yml:
```diff
@@ -14,6 +14,11 @@ en:
+        user_mentioned:
+          email_intro: You has been mentioned
+          email_outro: You have received this notification because you has been mentioned in "%{resource_title}".
+          email_subject: You has been mentioned in %{resource_title}
```
You have

mrcasals commented on decidim-comments/config/locales/en.yml:
```diff
@@ -14,6 +14,11 @@ en:
+          email_intro: You has been mentioned
+          email_outro: You have received this notification because you has been mentioned in "%{resource_title}".
+          email_subject: You has been mentioned in %{resource_title}
+          notification_title: You has been mentioned in <a href="%{resource_path}">%{resource_title}</a> by <a href="%{author_path}">%{author_name} %{author_nickname}</a>
```
You have

mrcasals commented on decidim-comments/lib/decidim/comments/api/comment_type.rb:
```diff
@@ -21,6 +21,8 @@ module Comments
 
       field :body, !types.String, "The comment message"
 
+      field :formattedBody, !types.String, "The comment message ready to display", property: :formatted_body
```
I think the description section should say that it is expected to include HTML

mrcasals commented on decidim-comments/spec/models/comment_spec.rb:
```diff
@@ -101,6 +101,14 @@ module Comments
+      describe "#formatted_body" do
+        let(:comment) { create(:comment, commentable: commentable, author: author, body: "<b>bold text</b> <a href='https://example.com'>link</a>") }
+
+        it "sanitizes user input striping html tags" do
```
s/striping/stripping/, s/html/HTML/

mrcasals replied:
Also, shouldn't this return HTML with links? :confused:

mrcasals commented on decidim-core/lib/decidim/content_parsers/base_parser.rb:
```diff
@@ -0,0 +1,29 @@
+
+      # Parse the the `content` and return it modified
+      def rewrite
+        raise NotImplementedError
```
Do not use `NotImplementedError`, as this is not its use case:

http://chrisstump.online/2016/03/23/stop-abusing-notimplementederror/

You cann `raise` with a string saying it should be defined, for example

rbngzlv replied:
It is always good to read articles like this, thank you. I'm refactoring to this:

```
def rewrite
  raise "#{self.class.name} is expected to implement #rewrite"
end
```

Also note that the `NotImplementedError` is used in other places of the code for this use case, so there is room to improve the base code. I do not have time right now to refactor the other parts. Perhaps we can open an issue as a `good first issue`, it's okay? Or can you take care of this?

mrcasals replied:
Open an issue please 😄

mrcasals commented on decidim-core/lib/decidim/content_parsers/user_parser.rb:
```diff
@@ -0,0 +1,24 @@
+module Decidim
+  module ContentParsers
+    class UserParser < BaseParser
+      MENTION_REGEX = /(^|\s)@([a-z0-9]\w*)/
```
Are underscores allowed as a valid nickname character?

rbngzlv replied:
Yes, the `\w` is a shortcut for `[a-zA-Z0-9_]`, but I will modify it a bit to also capture nicknames starting with a uppercase letter. So, the nickname can start with a letter or number, but not with a underscore. The final regex will be `(^|\s)@([a-zA-Z0-9]\w*)`

mrcasals commented on decidim-core/lib/decidim/content_parsers/user_parser.rb:
```diff
@@ -0,0 +1,24 @@
+
+module Decidim
+  module ContentParsers
+    class UserParser < BaseParser
```
Missing class docs. Please explain how to use this class

mrcasals commented on decidim-core/lib/decidim/content_processor.rb:
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module Decidim
+  module ContentProcessor
```
Missing docs (class & methods)

mrcasals commented on decidim-core/lib/decidim/core.rb:
```diff
@@ -93,6 +96,13 @@ def self.seed!
+  # Exposes a configuration option: an array of symbols that matches a ContentParser, and
+  # their equivalent ContentRenderer, that will be automatically executed when the content
+  # is parsed or rendered respectively.
+  config_accessor :content_processors do
```
Shouldn't this array should contain the `UserProcessor`? Also, missing docs on the expected syntax: how is this array supposed to be formatted? Is this an array of class names?

rbngzlv replied:
The `user` processor is registered in an initializer declared in the engine, as the `abilities`:

```
initializer "decidim.content_processors" do |_app|
  Decidim.configure do |config|
    config.content_processors += [:user]
  end
end
```

It's ok? Or better remove the initializer and add the processor directly here?

I'll add some examples of the expected syntax.

mrcasals commented on decidim-core/spec/content_parsers/decidim/user_parser_spec.rb:
```diff
@@ -0,0 +1,48 @@
+      let(:content) { "This text contains a valid user mention: @#{user.nickname}" }
+
+      it "rewrites the mention" do
+        expect(parser.rewrite).to include(user.to_global_id.to_s)
```
Can you check for the exact string to ensure the rest of the string is not modified?

`expect(...).to eq "This text contains a valid user mention: @#{user.to_global_id.to_s}"`

mrcasals commented on decidim-core/spec/content_renderers/decidim/user_renderer_spec.rb:
```diff
@@ -0,0 +1,39 @@
+      let(:content) { "This text contains a valid gid for user: #{user.to_global_id}" }
+
+      it "renders the mention" do
+        expect(render.render).to include(presenter.display_mention)
```
Can you check it returns the HTML link to the user?

mrcasals commented on decidim-core/spec/content_renderers/decidim/user_renderer_spec.rb:
```diff
@@ -0,0 +1,39 @@
+      end
+    end
+
+    context "when content has a invalid guid user" do
```
s/a invalid/an invalid/

mrcasals commented on decidim-core/spec/content_renderers/decidim/user_renderer_spec.rb:
```diff
@@ -0,0 +1,39 @@
+      end
+    end
+
+    context "when content has a unparsed mention" do
```
s/a unparsed/ an unparsed/

mrcasals commented on decidim-core/spec/content_renderers/decidim/user_renderer_spec.rb:
```diff
@@ -0,0 +1,39 @@
+      let(:content) { "This text contains a invalid gid for removed user: #{user.to_global_id.to_s.gsub(/\d$/, "0")}" }
+
+      it "removes the guid" do
+        expect(render.render).not_to include("guid:")
```
Shouldn't you be checking for `gid:` instead of `guid:`?

That's why I say it's better to check for the exact string in these cases

mrcasals commented on decidim-core/spec/content_renderers/decidim/user_renderer_spec.rb:
```diff
@@ -0,0 +1,39 @@
+        expect(render.render).not_to include("guid:")
+      end
+
+      it "not raises an exception" do
```
`it "does not raise an exception"`

mrcasals commented on decidim-dev/lib/decidim/dev/test/rspec_support/content_processing.rb:
```diff
@@ -0,0 +1,35 @@
+
+  class ContentRenderers::DummyFooRenderer < ContentRenderers::BaseRenderer
+    def render
+      content.gsub("_foo_", "foo")
```
Can you make it render to something else, so the initial content before parsing and the final one are different?

mrcasals replied:
Also for the `DummyBarRenderer`

tramuntanal commented:
I'm mentioning @tramuntanal here to stay updated with this PR because #2288 depends on this parser.

rbngzlv commented:
@mrcasals Thank you for your exhaustive review.

@decidim/lot-core ready for a second round? I added the last three commits.

mrcasals commented:
@rbngzlv there are conflicts, can you fix them before reviewing please?

rbngzlv commented:
@mrcasals Well, a pull request refactoring events was merged to master a few minutes after I pushed the changes. I rebased this branch onto master and added a commit to refactor the events.

oriolgual commented on decidim-comments/app/models/decidim/comments/comment.rb:
```diff
@@ -63,6 +63,11 @@ def reported_content_url
 
+      # Public: Returns the comment message ready to display (it is expected to include HTML)
+      def formatted_body
+        Decidim::ContentProcessor.render(sanitized_body)
```
Shouldn't we memoize this?

`@formatted_body ||= Decidim::ContentProcessor.render(sanitized_body)`

rbngzlv commented:
@mrcasals @oriolgual Thank you guys! All test passing now.

mrcasals commented:
Wohoooo! 🎉
